src/structures/MessageComponentInteractionCollector.js

```diff
@@ -0,0 +1,178 @@
+    });
+  }
+
+  /**
+   * Handles an incoming interaction for possible collection.
+   * @param {Interaction} interaction The interaction to possibly collect
+   * @returns {?Snowflake|string}
+   * @private
+   */
+  collect(interaction) {
+    /**
+     * Emitted whenever a interaction is collected.
+     * @event MessageComponentInteractionCollector#collect
+     * @param {Interaction} interaction The interaction that was collected
+     */
+    if (!interaction.isMessageComponent()) return null;
+
+    if (this.message) {
+      return interaction.message.id === this.message.id ? interaction.id : null;
+    }
+
+    return interaction.channel.id === this.channel.id ? interaction.id : null;
+  }
+
+  /**
+   * Handles an interaction for possible disposal.
+   * @param {Interaction} interaction The interaction that could be disposed of
+   * @returns {?Snowflake}
+   */
+  dispose(interaction) {
+    /**
+     * Emitted whenever an interaction is disposed of.
+     * @event MessageComponentInteractionCollector#dispose
+     * @param {Interaction} interaction The interaction that was disposed of
+     */
+    if (!interaction.isMessageComponent()) return null;
+
+    if (this.message) {
+      return interaction.message.id === this.message.id ? interaction.id : null;
+    }
+
+    return interaction.channel.id === this.channel.id ? interaction.id : null;
+  }
+
+  /**
+   * Empties this message component collector.
+   */
+  empty() {
+    this.total = 0;
+    this.collected.clear();
+    this.users.clear();
+    this.checkEnd();
+  }
+
+  get endReason() {
+    if (this.options.max && this.total >= this.options.max) return 'limit';
+    if (this.options.maxComponents && this.collected.size >= this.options.maxComponents) return 'componentLimit';
+    if (this.options.maxUsers && this.users.size >= this.options.maxUsers) return 'userLimit';
+    return null;
+  }
+
+  /**
+   * Handles checking if the message has been deleted, and if so, stops the collector with the reason 'messageDelete'.
+   * @private
+   * @param {Message} message The message that was deleted
+   * @returns {void}
+   */
+  _handleMessageDeletion(message) {
+    if (message.id === this.message.id) {
+      this.stop('messageDelete');
+    }
+  }
+
+  /**
+   * Handles checking if the channel has been deleted, and if so, stops the collector with the reason 'channelDelete'.
+   * @private
+   * @param {GuildChannel} channel The channel that was deleted
+   * @returns {void}
+   */
+  _handleChannelDeletion(channel) {
+    if (channel.id === this.message.channel.id) {
+      this.stop('channelDelete');
+    }
+  }
+
+  /**
+   * Handles checking if the guild has been deleted, and if so, stops the collector with the reason 'guildDelete'.
+   * @private
+   * @param {Guild} guild The guild that was deleted
+   * @returns {void}
+   */
+  _handleGuildDeletion(guild) {
+    if (this.message.guild && guild.id === this.message.guild.id) {
```

Comment:
This would throw a type error if a guild was deleted while a channel-scoped collector was running.
```suggestion
    if (guild.id === this.channel.guild?.id) {
```